Add keras-onnx code coverage report to CI

vinitra opened the pull request:
Adding benchmark.py with @xadupre's run_all_tests function to enable code coverage reports in the CI pipeline for keras-onnx.

jiafatom commented on .azure-pipelines/win32-conda-CI.yml:
```diff
@@ -62,6 +62,12 @@ jobs:
     displayName: 'pytest'
 
+  - script: |
+      coverage run --include=keras2onnx/** tests/benchmark.py
```
keras2onnx/ contains tf2onnx code, we should not consider coverage of them. Currently I am using 
coverage run --include=keras2onnx/ke2onnx/ **,keras2onnx/common/ **,keras2onnx/main.py,keras2onnx/funcbook.py,keras2onnx/parser.py,keras2onnx/subgraph.py,keras2onnx/topology.py,keras2onnx/wrapper.py,keras2onnx/_builtin.py tests/test_layers.py
to do the test.

vinitra replied:
Resolved in latest update.

jiafatom commented:
(1) Is there a way to write the coverage run more shortly? currently I listed all python files that is so long.
(2) for test/test_layers.py, I also want to add test_misc.py test_opset_change.py, applications/test_keras_applications.py, is there a way to do that?

jiafatom commented on .azure-pipelines/win32-conda-CI.yml:
```diff
@@ -62,6 +62,12 @@ jobs:
     displayName: 'pytest'
 
+  - script: |
+      coverage run --include=keras2onnx/** tests/** --omit=keras2onnx/ktf2onnx/**
```
seems that this script cannot work

vinitra replied:
Including only tests/*.py now! --omit should work after --include, so it will only include files not omitted in ktf2onnx/*

xadupre commented on .azure-pipelines/win32-conda-CI.yml:
```diff
@@ -62,6 +62,13 @@ jobs:
 
+  - script: |
+      pip install coverage onnx
+      coverage run --include=keras2onnx/* --omit=keras2onnx/ktf2onnx/* tests/test_layers.py
```
This command can be merged with pytest to run unit test and get the coverage at the same time. That might fix the build.

jiafatom commented:
I don't see the test coverage shown up on the screen on pytest section. We need fix that. Thanks!

vinitra commented:
@jiafatom, it exists at the bottom of the CI build for Win32 but the number looks lower than expected.
https://dev.azure.com/onnxmltools/ketone/_build/results?buildId=3283

```
Name                                   Stmts   Miss  Cover   Missing
--------------------------------------------------------------------
keras2onnx\__init__.py                    19      2    89%   24-25
keras2onnx\_builtin.py                   200    186     7%   18-20, 29-34, 38-155, 160-276, 280-291
keras2onnx\common\__init__.py              5      0   100%
keras2onnx\common\cvtfunc.py              18      2    89%   20-21
keras2onnx\common\data_types.py            1      0   100%
keras2onnx\common\interim.py             158     32    80%   172, 201, 206, 209, 229-230, 232-233, 236, 298, 306, 314-319, 335-338, 356, 374-377, 383-388, 426, 439
keras2onnx\common\intop.py                32      3    91%   53, 60, 64
keras2onnx\common\onnx_ops.py             35     26    26%   16-21, 24-48, 51-56, 59
keras2onnx\common\utils.py                38      0   100%
keras2onnx\funcbook.py                    16      5    69%   25-30
keras2onnx\ke2onnx\__init__.py             4      0   100%
keras2onnx\ke2onnx\activation.py          41     28    32%   34-66
keras2onnx\ke2onnx\adv_activation.py      30     25    17%   15-48
keras2onnx\ke2onnx\batch_norm.py          49     45     8%   12-79
keras2onnx\ke2onnx\bidirectional.py      161    151     6%   18-26, 32-285
keras2onnx\ke2onnx\common.py              24     21    12%   10-12, 16-35
keras2onnx\ke2onnx\conv.py               141    120    15%   22-35, 40-64, 69-182, 188-192, 196-197, 201-202, 206-207, 211-212, 216-217, 221-222, 226-227, 231-232
keras2onnx\ke2onnx\crop.py                37     31    16%   11-50, 54, 58, 62
keras2onnx\ke2onnx\dense.py               22      1    95%   41
keras2onnx\ke2onnx\dot.py                 17     15    12%   10-28
keras2onnx\ke2onnx\embedding.py           17     13    24%   13-35
keras2onnx\ke2onnx\gru.py                 81     76     6%   13-121
keras2onnx\ke2onnx\lstm.py               107     98     8%   16-21, 26-168
keras2onnx\ke2onnx\main.py                53     12    77%   33, 37, 41-44, 49, 58-64
keras2onnx\ke2onnx\pooling.py             50     42    16%   11-54, 59-60, 65-66, 71-72, 77-78, 83-84, 89-90
keras2onnx\ke2onnx\simplernn.py           67     62     7%   13-100
keras2onnx\ke2onnx\upsample.py            31     24    23%   12-50, 54, 58, 62
keras2onnx\main.py                        54      5    91%   28, 54-55, 79, 87
keras2onnx\parser.py                     367    105    71%   24, 26, 36, 38, 41-46, 65, 75-87, 93, 98, 102-107, 112-163, 194, 232, 266, 290, 299, 338, 340, 351, 353, 360-384, 448, 462, 471
keras2onnx\proto\__init__.py              21      4    81%   10, 15-16, 37
keras2onnx\subgraph.py                    93     24    74%   21, 28, 31, 62-68, 70, 86-95, 101-106, 110, 121
keras2onnx\topology.py                   167     22    87%   81, 95, 123, 142, 194, 199, 210-212, 217, 225, 242, 254-262, 269, 294
keras2onnx\wrapper.py                     48     13    73%   26-29, 54-60, 67-70, 77-78
--------------------------------------------------------------------
TOTAL                                   2204   1193    46%
##[section]Finishing: pytest + coverage

```

jiafatom commented:
@vinitra This is probably because coverage only picks the latest command. Let's just do tests/test_layers.py only.

vinitra commented:
Updated! The 89% test coverage is now apparent. I've also updated the windows nightly pipeline with coverage.

```
Name                                   Stmts   Miss  Cover   Missing
--------------------------------------------------------------------
keras2onnx\__init__.py                    19      2    89%   24-25
keras2onnx\_builtin.py                   200     10    95%   33-34, 42, 91-92, 164, 185, 200, 219-220
keras2onnx\common\__init__.py              5      0   100%
keras2onnx\common\cvtfunc.py              18      2    89%   20-21
keras2onnx\common\data_types.py            1      0   100%
keras2onnx\common\interim.py             158     31    80%   172, 201, 209, 229-230, 232-233, 236, 298, 306, 314-319, 335-338, 356, 374-377, 383-388, 426, 439
keras2onnx\common\intop.py                32      2    94%   60, 64
keras2onnx\common\onnx_ops.py             35     26    26%   16-21, 24-48, 51-56, 59
keras2onnx\common\utils.py                38      0   100%
keras2onnx\funcbook.py                    16      5    69%   25-30
keras2onnx\ke2onnx\__init__.py             4      0   100%
keras2onnx\ke2onnx\activation.py          41      1    98%   64
keras2onnx\ke2onnx\adv_activation.py      30     12    60%   33-48
keras2onnx\ke2onnx\batch_norm.py          49      3    94%   15-18
keras2onnx\ke2onnx\bidirectional.py      161     15    91%   41, 43, 111, 143, 190-191, 197, 200, 274-285
keras2onnx\ke2onnx\common.py              24      5    79%   20, 29-33
keras2onnx\ke2onnx\conv.py               141      7    95%   139-142, 147, 157, 221-222
keras2onnx\ke2onnx\crop.py                37      8    78%   21, 28, 32, 41-43, 50, 54, 62
keras2onnx\ke2onnx\dense.py               22      0   100%
keras2onnx\ke2onnx\dot.py                 17      2    88%   12, 14
keras2onnx\ke2onnx\embedding.py           17      1    94%   15
keras2onnx\ke2onnx\gru.py                 81      6    93%   51, 88-89, 91-92, 121
keras2onnx\ke2onnx\lstm.py               107      9    92%   96-99, 105-108, 139-140, 162
keras2onnx\ke2onnx\main.py                53      3    94%   33, 37, 64
keras2onnx\ke2onnx\pooling.py             50      4    92%   15, 33, 39, 43
keras2onnx\ke2onnx\simplernn.py           67      6    91%   48, 66, 68, 75-76, 97
keras2onnx\ke2onnx\upsample.py            31      6    81%   15, 21, 25, 37, 46, 54
keras2onnx\main.py                        54      4    93%   28, 54-55, 87
keras2onnx\parser.py                     367     31    92%   24, 26, 36, 38, 41-46, 75-87, 98, 102-107, 161, 232, 290, 351, 462
keras2onnx\proto\__init__.py              21      3    86%   10, 15-16
keras2onnx\subgraph.py                    93     11    88%   21, 28, 31, 66, 101-106, 110
keras2onnx\topology.py                   167     17    90%   81, 95, 123, 142, 194, 199, 217, 225, 242, 254-262
keras2onnx\wrapper.py                     48      7    85%   26-29, 54-60
--------------------------------------------------------------------
TOTAL                                   2204    239    89%
```

jiafatom commented:
can we do it in Linux as well?
